### PR TITLE
Presidio power skill

### PR DESCRIPTION
**Pull request by shiranr**
## Purpose
Add new Power Skill to the repository. PII anonymization using Presidio.

## Does this introduce a breaking change?
```
[ ] Yes
[ X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?
```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone https://github.com/shiranr/azure-search-power-skills.git
cd azure-search-power-skills.git
git checkout presidio_power_skill
npm install
```

* Test the code
The readme contains all the instructions including an updated notebook 

## What to Check
This Power Skill works with PythonFastAPI and built with the same logic as the text summarization. 
The result of the indexing should be anonymized text, all PII entities should be removed from the text.

### CODE REVIEW

**Comment by SharonHart** [comment]
Why aren't we using the official presidio docker? the API layer here is not immune to changes, and doesn't contain lot's of the stuff we did with error handling

File: Text/PIIAnonymization/README.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
+# Text PII Anonymization Power Skill
+
+This Power Skill uses [Presidio](https://github.com/microsoft/presidio) Analyzer and Anonymizer
+to find and remove PII entities. Even though Presidio supports several anonymization methods (hash, encrypt, redact, replace, mask),
```
Comment (navalev) [inline Text/PIIAnonymization/README.md]: Lets change the order of the sentances here - put this one after explaining about what presidio is

Comment (shiranr) [reply Text/PIIAnonymization/README.md]: Done. Also renamed the package :) and added better text to the indexer.

File: README.md [hunk 2]
```diff
@@ -20,6 +20,7 @@ This project provides the following custom skills:
 | [Distinct](Text/Distinct/README.md) | de-duplicates a list of terms. | Text | ![C#](https://img.shields.io/badge/language-C%23-brightgreen) | ![functions](https://img.shields.io/badge/deploy-Functions-blue) | ARM Template |
 | [Tokenizer](Text/Tokenizer/README.md) | extracts non-stop words from a text. | Text | ![C#](https://img.shields.io/badge/language-C%23-brightgreen) | ![functions](https://img.shields.io/badge/deploy-Functions-blue) |
 | [Summarizer](Text/TextSummarization/README.md) | Uses a HuggingFace/FaceBook BART model to summarize text [BART-Large-CNN](https://huggingface.co/facebook/bart-large-cnn). | Text | ![python](https://img.shields.io/badge/language-python-orange) | ![docker](https://img.shields.io/badge/deploy-Docker-blueviolet) | Manual |
+| [Anonymizer](Text/PIIAnonymization/README.md)  | Uses [Presidio](https://github.com/microsoft/presidio) to analyze and anonymize PII entities.  | Text | ![python](https://img.shields.io/badge/language-python-orange) | ![docker](https://img.shields.io/badge/deploy-Docker-blueviolet) | Manual |
```
Comment (omri374) [inline README.md]:
```suggestion
| [Anonymizer](Text/PresidioPIIAnonymization/README.md)  | Uses [Presidio](https://github.com/microsoft/presidio) to analyze and anonymize PII entities.  | Text | ![python](https://img.shields.io/badge/language-python-orange) | ![docker](https://img.shields.io/badge/deploy-Docker-blueviolet) | Manual |
```

File: Text/PresidioPIIAnonymization/README.md [hunk 3]
```diff
@@ -0,0 +1,215 @@
+# Text PII Anonymization Power Skill
+
+[Presidio](https://github.com/microsoft/presidio) is an open-source tool to recognize, analyze and anonymize personally identifiable information (PII). Using trained ML models, Presidio was built to ensure sensitive text is properly managed and governed.
+
+This Power Skill uses [Presidio](https://github.com/microsoft/presidio) Analyzer and Anonymizer
+to find and remove PII entities. Even though Presidio supports several anonymization methods (hash, encrypt, redact, replace, mask), 
+this Power Skill only uses redact and removes the PIIs completely from the text  .
+
+This skill is ideal for finding and removing PII entities from the search text.
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md:9]: Consider adding a disclaimer like we have on our readme, that we cannot guarantee 100% accuracy

File: Text/PresidioPIIAnonymization/README.md [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+## Requirements
+
+In addition to the common requirements described in the root [README.md](https://github.com/Azure-Samples/azure-search-power-skills/blob/master/README.md) 
+file, this Power Skill requires spacy en_core_web_lg module being downloaded:
+```python
+python -m spacy download en_core_web_lg
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md:16]: Is this the only requirement? What about presidio-analyzer and presidio-anonymizer? 
Also, it's not clear if these are the requirements to develop locally or to deploy the power skill into the user's cognitive search pipeline.

File: Text/PresidioPIIAnonymization/README.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+```
+
+To run this PowerSkill you will need:
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md:19]: Consider having this prior to the spacy stuff, as this is the basics for setting up a power skill

File: Text/PresidioPIIAnonymization/README.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+* docker
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md]:
```suggestion
* Docker
```

File: Text/PresidioPIIAnonymization/README.md [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+* A Cognitive Services key in the region you deploy ACS to
+
+Below is a full working example that you can get working E2E on sample data.
+
+## How to implement
+
+This section describes how to get this working on sample data and how it can be amended for your data.
+ 
+1) ### Data
+   The first step is to view the sample data files here [train data](data/).
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md]:
```suggestion
   The first step is to view the sample data. [Link to sample data](data/). 
```

File: Text/PresidioPIIAnonymization/README.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+This section describes how to get this working on sample data and how it can be amended for your data.
+ 
+1) ### Data
+   The first step is to view the sample data files here [train data](data/). 
+1) ### Run the API
+   The next step is to run the API locally and test the model against a test record. Create a local python environment
+   and install the requirements:
+   
+   ```python
+   python -m pip install -r requirements.txt
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md]:
```suggestion
   python -m pip install -r powerskill/requirements.txt
```

File: Text/PresidioPIIAnonymization/README.md [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+Below is a full working example that you can get working E2E on sample data.
+
+## How to implement
+
+This section describes how to get this working on sample data and how it can be amended for your data.
+ 
+1) ### Data
+   The first step is to view the sample data files here [train data](data/). 
+1) ### Run the API
+   The next step is to run the API locally and test the model against a test record. Create a local python environment
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md:31]: If by API you mean the one that's exposed in this power skill, worth mentioning explicitly.

File: Text/PresidioPIIAnonymization/README.md [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+   Run the following command to build the inference API container image:
+
+    ```bash
+    docker build -t [container_registry_name.azurecr.io]/pii_anonymization:[your_tag] .  
+    ```
+    
+    The container will require the following variables set at runtime, namely:
+    
+    ```bash
+    KEY=[YourSecretKeyCanBeAnything]    # This is a secret key - only requests with this key will be allowed
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md:66]: Is this mandatory for any powerskill? I'm not sure what this key does and how to use it.

File: Text/PresidioPIIAnonymization/README.md [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+    running a container.
+    
+    First we need to push our newly built image to our container registry.
+    
+    Run the following command:
+    ```bash
+    docker push [container_registry_name].azurecr.io/pii_anonymization:[your_tag]
+    ```
+    
+    In the [deployment folder](deployment/webapp) are two [terraform](https://www.terraform.io/)
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md]:
```suggestion
    In the [deployment folder](deployment/webapp) there are two [terraform](https://www.terraform.io/)
```

File: Text/PresidioPIIAnonymization/README.md [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    ```hcl-terraform
+    backend "azurerm" {
+        storage_account_name = "[your storage account name]"
+        container_name = "[your storage container name]"
+        key = "[your storage account key]"
+        resource_group_name = "[your storage account resource group name]"
+      }
+    ```
+    
+    Set the following values in the[variables](deployment/webapp/variables.tf)
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md]:
```suggestion
    Set the following values in the [variables](deployment/webapp/variables.tf)
```

File: Text/PresidioPIIAnonymization/README.md [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+   ```python 
+   python app.py
+   ``` 
+   Run the cell 
+   [Test PII anonymization on our local running API](notebooks/PII%20Anonymization.ipynb#Test-our-text-on-our-local-running-API). 
+   Make sure you rename the file [sample_env file to .env](powerskill/sample_env) and populate it with the relevant values. Use the
+   variable ```bash URL_LOCAL``` as the URL.
+1) ### Build the docker image 
+   Now build the [docker image](Dockerfile) and upload the image to your container registry  
+   For this step you will need docker running so that we can build and test our inference API locally.
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md]: Inference API can be confusing. Consider being more explicit and call it "Anonymization API" or "Presidio inference API" or something similar.

File: Text/PresidioPIIAnonymization/notebooks/.ipynb_checkpoints/PII Anonymization-checkpoint.ipynb [hunk 14]
```diff
@@ -0,0 +1,381 @@
+{
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/notebooks/.ipynb_checkpoints/PII Anonymization-checkpoint.ipynb]: This file should be removed, it's just a jupyter checkpoint file

File: Text/PresidioPIIAnonymization/notebooks/PII Anonymization.ipynb [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+    "# Test our text on our local running API"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "API_KEY = \"[YourSecretKeyCanBeAnything]\"\n",
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/notebooks/PII Anonymization.ipynb]: Consider adding more information on what this key is and how to use/obtain it. (Maybe it's my limited knowledge of PowerSkills.)

File: Text/PresidioPIIAnonymization/powerskill/app.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+from fastapi.security.api_key import APIKeyHeader, APIKey
+from objdict import ObjDict
+from pydantic import BaseModel
+from starlette.status import HTTP_400_BAD_REQUEST, HTTP_403_FORBIDDEN
+
+from powerskill import Presidio
+
+load_dotenv()
+app = FastAPI()
+presidio = Presidio()
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/powerskill/app.py:15]: nit: Why not use the AnalyzerEngine and AnonymizerEngine directly? Might be easier for users to then customize

File: Text/PresidioPIIAnonymization/README.md (hunk 8 above)
Comment (omri374) [reply Text/PresidioPIIAnonymization/README.md]: Why not add the spacy model installation here?

File: Text/PresidioPIIAnonymization/README.md [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+
+Below is a full working example that you can get working E2E on sample data.
+
+## How to implement
+
+This section describes how to get this working on sample data and how it can be amended for your data.
+ 
+1) ### Data
+   The first step is to view the sample data files here [train data](data/). 
+1) ### Run the API
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md]: Consider splitting into Install and then Run

File: Text/PresidioPIIAnonymization/README.md [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+
+[Presidio](https://github.com/microsoft/presidio) is an open-source tool to recognize, analyze and anonymize personally identifiable information (PII). Using trained ML models, Presidio was built to ensure sensitive text is properly managed and governed.
+
+This Power Skill uses [Presidio](https://github.com/microsoft/presidio) Analyzer and Anonymizer
+to find and remove PII entities. Even though Presidio supports several anonymization methods (hash, encrypt, redact, replace, mask), 
+this Power Skill only uses redact and removes the PIIs completely from the text  .
+
+This skill is ideal for finding and removing PII entities from the search text.
+
+## Requirements
```
Comment (omri374) [inline Text/PresidioPIIAnonymization/README.md]: Are these requirements for installing locally or deploying?

File: Text/PresidioPIIAnonymization/README.md (hunk 9 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/README.md:31]: Done.

File: Text/PresidioPIIAnonymization/README.md (hunk 3 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/README.md:9]: Done.

File: Text/PresidioPIIAnonymization/README.md (hunk 13 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/README.md]: Done.

File: Text/PresidioPIIAnonymization/notebooks/PII Anonymization.ipynb (hunk 15 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/notebooks/PII Anonymization.ipynb]: It's copy paste from the original redame.

File: Text/PresidioPIIAnonymization/notebooks/.ipynb_checkpoints/PII Anonymization-checkpoint.ipynb (hunk 14 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/notebooks/.ipynb_checkpoints/PII Anonymization-checkpoint.ipynb]: In the text summarization it is submitted. But I don't mind removing it.

File: Text/PresidioPIIAnonymization/notebooks/PII Anonymization.ipynb (hunk 15 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/notebooks/PII Anonymization.ipynb]: There is an explanation somewhere else there... I copied it.

File: Text/PresidioPIIAnonymization/powerskill/app.py (hunk 16 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/powerskill/app.py:15]: It is more clean and clear IMO. Also, it is replica of the behavior separation which was done in the summary.

File: Text/PresidioPIIAnonymization/README.md (hunk 17 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/README.md]: I followed the summarization way of putting it down.

File: Text/PresidioPIIAnonymization/README.md (hunk 18 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/README.md]: Done.

File: Text/PresidioPIIAnonymization/README.md (hunk 10 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/README.md:66]: it's part of the authentication for the application. It is done in other powerskills as well, yes.

File: Text/PresidioPIIAnonymization/README.md (hunk 11 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/README.md]: Done.

File: Text/PresidioPIIAnonymization/README.md (hunk 12 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/README.md]: Done.

File: Text/PresidioPIIAnonymization/powerskill/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,19 @@
+FROM python:3.8
+
+# [Optional] If your pip requirements rarely change, uncomment this section to add them to the image.
```
Comment (SharonHart) [inline Text/PresidioPIIAnonymization/powerskill/Dockerfile]: I might be missing something ,but nothing is commented-out?

File: Text/PresidioPIIAnonymization/powerskill/Dockerfile_debug [hunk 20]
```diff
@@ -0,0 +1,25 @@
+FROM python:3.8
+
+# [Optional] If your pip requirements rarely change, uncomment this section to add them to the image.
```
Comment (SharonHart) [inline Text/PresidioPIIAnonymization/powerskill/Dockerfile_debug]: here as well

File: Text/PresidioPIIAnonymization/powerskill/requirements.txt [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+azure-core==1.12.0
+azure-search==1.0.0b2
+azure-search-documents==11.1.0
+presidio-analyzer==2.2.0
+presidio-anonymizer==2.2.0
+pydantic~=1.7.3
+starlette~=0.13.6
+requests~=2.25.1
+numpy~=1.20.2
+scipy~=1.6.2
```
Comment (SharonHart) [inline Text/PresidioPIIAnonymization/powerskill/requirements.txt]: I don't see any usage in some of the stuff, like numpy, scipy. Why are those needed?

**Review by SharonHart — APPROVED** [review]
LGTM! left some questions

**Review by omri374 — APPROVED** [review]

**Comment by omri374** [comment]
Looks great! awesome work

File: Text/PresidioPIIAnonymization/powerskill/Dockerfile (hunk 19 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/powerskill/Dockerfile]: Done.

File: Text/PresidioPIIAnonymization/powerskill/Dockerfile_debug (hunk 20 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/powerskill/Dockerfile_debug]: Done.

File: Text/PresidioPIIAnonymization/powerskill/requirements.txt (hunk 21 above)
Comment (shiranr) [reply Text/PresidioPIIAnonymization/powerskill/requirements.txt]: it was automatically added... removed and tested. Also fixed the versions.
Good comments. thank you.
